datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -171,6 +171,28 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
     return super.assign(datastreams, instances, currentAssignment);
   }
 
+  protected List<String> getAssignedPartitionsForDatastreamGroup(Map<String, Set<DatastreamTask>> currentAssignment,
```

Comment:
Would it have worked, if you would have returned `Pair`. This change is directly adding an additional iteration to the existing code as well.

Also, for the existing strategy, it will walk the tasks list twice now and for new strategy, it will walk 3 times. There is a scope to extract out another method that gets assignedPartitions and taskcount as field and the new strategy can call it to avoid extra iteration.

See if you can address the `Pair` one, so it is exactly same as current code and you can mark TODO to reduce additional iteration for new strategy.

Reply:
It's a tradeoff between keeping the code readable and avoiding extra iteration. I can make the method calculate both assigned partitions and current task count and return a pair if you think it's better that way.